### PR TITLE
Add dynamic TLS Configuration for the streaming server

### PR DESCRIPTION
**Pull request by nathanjsweet**
Signed-off-by: Nathan Sweet <nathanjsweet@gmail.com>

fixes: #1582 

**- What I did**
Add Dyanmic TLS Configuration for the streaming server. Optional CA verification is added, though kubernetes doesn't seem to be requesting the streaming server with a client cert, the option is there to add it.

**- How I did it**
I simply added a TLSConfig object to stream server config object, that implements GetConfigForClient method.
**- How to verify it**
Start crio, add the `tls_streaming`, `tls_cert`, and `tls_key` options, using the kubelet cert and key for testing. The kube-apiserver also needs to have the CA set that also signed these certs.
**- Description for the changelog**
Added TLS Encryption to the streaming server

### CODE REVIEW

File: cmd/crio/config.go [hunk 1]
```diff
@@ -46,6 +46,19 @@ stream_address = "{{ .StreamAddress }}"
 # stream_port is the port on which the stream server will listen
 stream_port = "{{ .StreamPort }}"
 
+# tls_streaming enables encrypted tls transport of the stream server
+tls_streaming = {{ .TLSStreaming }}
```
Comment (mrunalp) [inline cmd/crio/config.go]: Can we name this `stream_enable_tls` ?

File: cmd/crio/config.go [hunk 2]
```diff
@@ -46,6 +46,19 @@ stream_address = "{{ .StreamAddress }}"
...
+
+# tls_cert is the x509 certificate file path used to serve the encrypted stream
+tls_cert = "{{ .TLSCert }}"
```
Comment (mrunalp) [inline cmd/crio/config.go]: Also prepend all of these with `stream_` ?

**Comment by mrunalp** [comment]
Can you fixup the commit title? One of the validations is failing with whitespace errors.

**Comment by TomSweeneyRedHat** [comment]
Changes LGTM, anyway to work up some tests for the new code?

**Comment by nathanjsweet** [comment]
I'm down to write tests, can you direct me to the streaming tests?

**Comment by mrunalp** [comment]
@nathanjsweet Thanks! For the tests, I think they should be contributed to critest in the cri-tools repository so it helps all the runtimes. They aren't a blocker for this PR though.

**Comment by nathanjsweet** [comment]
I'm not really understanding this RHEL failure. It doesn't look like something we could have caused. Or is there something I'm missing?

File: server/server.go [hunk 3]
```diff
@@ -326,6 +370,24 @@ func New(ctx context.Context, config *Config) (*Server, error) {
 	// Prepare streaming server
 	streamServerConfig := streaming.DefaultConfig
 	streamServerConfig.Addr = net.JoinHostPort(bindAddress.String(), config.StreamPort)
+	if config.StreamEnableTLS {
+		certCache = &certConfigCache{
```
[the remark this answers is not in the document]
Comment (nathanjsweet) [reply server/server.go]: I was under the impression `New()` was only called the one time to set up the servers, but I guess that's not the case. You say `New()` is called concurrently? I guess I'll throw a mutex in as well (if we stick with this rotation thing).

File: server/server.go [hunk 4]
```diff
@@ -75,6 +83,42 @@ type Server struct {
...
+	config.Certificates = []tls.Certificate{cert}
+	if len(cc.tlsCA) > 0 {
+		caBytes, err := ioutil.ReadFile(cc.tlsCA)
+		if err != nil {
+			return nil, err
+		}
+		certPool := x509.NewCertPool()
+		certPool.AppendCertsFromPEM(caBytes)
+		config.RootCAs = certPool
+		config.ClientAuth = tls.RequireAndVerifyClientCert
```
[the remark this answers is not in the document]
Comment (nathanjsweet) [reply server/server.go]: I can look into where the tests should go. I will double check, but if we absolutely need tests it does not look like the kubernetes apiserver is making the request with a certificate. I will confirm this. If it is the case then I will remove this portion of the PR for now, and leave a comment.

File: server/server.go [hunk 5]
```diff
@@ -75,6 +83,42 @@ type Server struct {
...
+	expires time.Time
+
+	tlsCert string
+	tlsKey  string
+	tlsCA   string
+}
+
+// GetConfigForClient gets the tlsConfig for the streaming server.
+// This allows the certs to be swapped, without shutting down crio.
+func (cc *certConfigCache) GetConfigForClient(hello *tls.ClientHelloInfo) (*tls.Config, error) {
```
[the remark this answers is not in the document]
Comment (nathanjsweet) [reply server/server.go]: I don't think there is a great story for certificate rotation in Kubernetes right now. We are doing this, because we come from a hyper secure posture of dealing with healthcare compliance and we have an aggressively attenuated schedule for expiring our PKI leaf certificates. Given that is the case, it seems overly inconvenient to have to stop CRIO just to reload a certificate.

Netflix actually expires their server certificates [every 4 days](https://medium.facilelogin.com/short-lived-certificates-netflix-fd5f3ae5bc9) (or at least has proposed it as a security feature).

File: server/server.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mtrmac) [reply server/server.go]: I’m not at all saying that `New` definitely is called concurrently; but if the code is structured to keep all state in the returned `Server` object, why break the st

[the remark this answers is not in the document]
Comment (mtrmac) [reply server/server.go]: First of all, what is the global state _supposed to do_? Why does it exist at all?

I’m not at all saying that `New` definitely is called concurrently; but if the code is structured to keep all state in the returned `Server` object, it seems preferable not to break that property without a good reason.

I can’t see why a mutex is necessary at all; just use a local variable here, `certCache := &certConfigCache{}` and everything will work.

If, as you say, there is always a single instance, worrying about mutexes and possibly sharing the cache is unnecessary work — and if there actually _were_ multiple concurrent instances, they could differ in the TLS settings inside `config`, so a single cache instance shared between all of them wouldn’t be appropriate anyway without much more infrastructure.

**Comment by nathanjsweet** [comment]
@mtrmac You've raised good points. However, it looks like the AuthZ RFC has been neglected for about a year and a half, but the main thrust of the design is that it will utilize the existing cluster PKI (similar to how the kube-apiserver talks to the kubelet). In terms of figuring out tests, it is probably not possible since the extension to the interface hasn't been implemented yet, even if I created a PR for the streaming tests in the kubernetes tests repo they would likely get rejected as other container runtimes would fail the tests.

I'll confirm this for sure, but I think the best way forward is to remove the ClientCAs check if the kube-apiserver is indeed not requesting with a cert, and content ourselves with the knowledge that we are at least mimicking part of the PKI Auth structure of the kubelet-to-kube-apiserver relationship.

File: server/server.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (nathanjsweet) [reply server/server.go]: > I can’t see why a mutex is necessary at all; just use a local variable here, certCache := &certConfigCache{} and everything will work.

I'm a little worried about the garbage collector given that there isn't a direct reference to the variable, which is why I made it global.

[the remark this answers is not in the document]
Comment (mtrmac) [reply server/server.go]: - `GetConfigForClient: certCache.GetConfigForClient` is enough of a reference.
- (Even if it were necessary to keep a reference, it should conceptually be attached to the lifetime of the `Server` object, not kept around forever.)

**Comment by jcarley** [comment]
Are these failures similar to the ones before where we just had to kick the test runs again?

**Comment by nathanjsweet** [comment]
@mrunalp It doesn't look like the test reran. I can confirm that these latest changes are working fine in our cluster.

**Comment by mrunalp** [comment]
Restarted flake.

**Comment by nathanjsweet** [comment]
@runcom @mrunalp I feel good about the state of this PR. Until the K8s AuthZ interface for the streaming service is implemented I legitimately don't see a way forwards on testing the mutual authentication piece. 

However, just having the encrypted streaming piece can certainly be tested and really helps those of us with compliance burdens (healthcare and finance), and I am happy to help with those tests.

I dropped some notes on [Tim (St. Clair) Allclair's design doc](https://docs.google.com/document/d/1OE_QoInPlVCK9rMAx9aybRmgFiVjHpJCHI9LrfdNM_s/edit#heading=h.4yfjiw58o8d3) for this and suggested a couple of paths forward to go one step further than authentication (this PR) and add authorization in as well (RBAC, etc).

Let me know if there is anything else you would like from me.

Thanks for all your help!

**Comment by mrunalp** [comment]
@nathanjsweet Thanks! @mtrmac are we okay with merging this?

**Comment by nathanjsweet** [comment]
@mtrmac If you take a look at the discussion that Tim and I had on that streaming design doc you'll see that right now exec, port forward, and attach are all redirected directly to the kube-apiserver if th Container Runtime is remote (like CRIO), eventually all traffic will flow exclusively through the kubelet, mitigating any need for the streaming server to be remotely accessible. 

I could still see this being useful to ensure that only the kubelet (as a process on the same host) can communicate with the container runtime.

To sum up:
1. This PR is useful now and keeps the stream encrypted between hosts, it does not address authorization issues though.
2. This code will become less useful once kubernetes removes the redirect from streaming/forwarding and proxies everything through the kubelet, but it will still be moderately useful to protect the streaming server from other system processes.

**Comment by mtrmac** [comment]
> @mtrmac are we okay with merging this?

The implementation LGTM. I don’t know about how the certs get managed, I’ll happily defer that to subject matter experts.

The way I read the above, if @nathanjsweet finds the PR useful right now, fine; the only thing that _needs_ to be considered is that if eventually the kubelet will locally proxy the connection and the kubelet <-> streaming server connection will be established / authenticated in some possibly completely different way, maybe making the config options introduced in this PR irrelevant or completely ignored. If that is fine with CRI-O, good.

**Comment by mrunalp** [comment]
Thanks! We are fine merging this for now.

**Comment by nathanjsweet** [comment]
@mrunalp @mtrmac Thanks for all your help!
